targets/TARGET_Freescale/TARGET_MCUXpresso_MCUS/api/flash_api.c

```diff
@@ -0,0 +1,126 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "mbed_assert.h"
+#include "flash_api.h"
+#include "mbed_critical.h"
+
+#if DEVICE_FLASH
+
+#include "fsl_flash.h"
+
+// Storage for flash driver.
+flash_config_t g_flash;
+
+int32_t flash_init(flash_t *obj)
+{
+    status_t result;
+
+    /* Clean up Flash driver Structure*/
+    memset(&g_flash, 0, sizeof(flash_config_t));
+
+    /* Setup flash driver structure for device and initialize variables. */
+    result = FLASH_Init(&g_flash);
+    if (kStatus_FLASH_Success != result) {
+        return -1;
+    } else {
+        return 0;
+    }
+}
+
+int32_t flash_free(flash_t *obj)
+{
+    return 0;
+}
+
+int32_t flash_erase_sector(flash_t *obj, uint32_t address)
+{
+    int status;
+
+    core_util_critical_section_enter();
+    status = FLASH_Erase(&g_flash, address, g_flash.PFlashSectorSize, kFLASH_ApiEraseKey);
+
+    if (status == kStatus_Success) {
+        status = FLASH_VerifyErase(&g_flash, address, g_flash.PFlashSectorSize, kFLASH_MarginValueNormal);
+    }
+    core_util_critical_section_exit();
+
+    if (status == kStatus_Success) {
+        return 0;
+    } else {
+        return -1;
+    }
+}
+
+int32_t flash_program_page(flash_t *obj, uint32_t address, const uint8_t *data, uint32_t size)
+{
+    int status;
+
+    core_util_critical_section_enter();
+    status = FLASH_Program(&g_flash, address, (uint32_t *)data, size);
+
+    if (status == kStatus_Success) {
+        // Must use kFlashMargin_User, or kFlashMargin_Factory for verify program
+        status = FLASH_VerifyProgram(&g_flash, address, size,
+                              (uint32_t *)data, kFLASH_MarginValueUser,
+                              NULL, NULL);
+    }
+    core_util_critical_section_exit();
+
+    return status;
+
+}
+
+uint32_t flash_get_sector_size(const flash_t *obj, uint32_t address)
+{
+    if ((address < FSL_FEATURE_FLASH_PFLASH_START_ADDRESS) ||
+        (address >= (FSL_FEATURE_FLASH_PFLASH_START_ADDRESS + FSL_FEATURE_FLASH_PFLASH_BLOCK_SIZE))) {
+        return MBED_FLASH_INVALID_SIZE;
+    } else {
+        return FSL_FEATURE_FLASH_PFLASH_BLOCK_SECTOR_SIZE;
+    }
+}
+
+uint32_t flash_get_page_size(const flash_t *obj)
+{
+    return 1024;
```

Comment:
Minimum write unit is 8 bytes. Is this what is expected for Page size. This would slow down the program. 1024 should work.